### PR TITLE
Creates the Error Code Framework and updates type codes.

### PR DESCRIPTION
**Pull request by abel2-code**
**Reason for Change**:
The goal of these changes is to enable the CNI client modules to receive consistent and meaningful feedback from CNS when an operation fails.  This will improve the telemetry and retry logic of the CNI client modules and enhance the reliability and performance of the container networking services.

**Issue Fixed**:
N/A

**Requirements**:
- [ ] uses [conventional commit messages]
- [X] includes documentation
- [X] adds unit tests
- [ ] relevant PR labels added

**Notes**:
This is part of an effort to create an extensible error code framework for CNS.

### CODE REVIEW

File: cns/types/codes.go [hunk 1]
```diff
@@ -3,47 +3,53 @@ package types
...
-	NetworkContainerVfpProgramCheckSkipped ResponseCode = 36
-	NmAgentSupportedApisError              ResponseCode = 37
-	UnsupportedNCVersion                   ResponseCode = 38
-	FailedToRunIPTableCmd                  ResponseCode = 39
-	NilEndpointStateStore                  ResponseCode = 40
-	NmAgentInternalServerError             ResponseCode = 41
-	StatusUnauthorized                     ResponseCode = 42
-	UnsupportedAPI                         ResponseCode = 43
-	UnexpectedError                        ResponseCode = 99
+	Success                                ResponseCode = 2000
```
Comment (ramiro-gamarra) [inline cns/types/codes.go:13]: These are values that may already be in use by CNS API clients. As arbitrary as they seem, they cannot be changed with any existing CNS API version.

**Review by ramiro-gamarra — CHANGES_REQUESTED** [review]
It would be good to understand the goal with these changes. I like the idea of coherent error definitions but they may only be added in a new version of CNS.

File: cns/types/codes.go (hunk 1 above)
Comment (abel2-code) [reply cns/types/codes.go:13]: I've updated the reason for the change to better explain the purpose of these changes.

**Comment by abel2-code** [comment]
@microsoft-github-policy-service agree [company="Microsoft"]

File: cns/types/codes.go (hunk 1 above)
Comment (jpayne3506) [reply cns/types/codes.go]: Welcome @abel2-code, please take a look at https://github.com/Azure/azure-container-networking/blob/master/docs/contributing/github.md when you get the chance. 

For now, lets allow @ramiro-gamarra to close the conversation when he feels it is resolved. ✌️

**Comment by ramiro-gamarra** [comment]
Just as an FYI, we do have Azure API guidelines which we could consider adopting/learning from going forward. Specific guidance on [errors](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#handling-errors).

File: cns/errorcodeframework/codes.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	ReservationNotFound                    ResponseCode = 4141
+	UnknownContainerID                     ResponseCode = 4151
+	UnreachableDockerDaemon                ResponseCode = 4161
+	UnreachableHost                        ResponseCode = 4164
+	UnspecifiedNetworkName                 ResponseCode = 4175
+	UnsupportedEnvironment                 ResponseCode = 4181
+	UnsupportedNetworkType                 ResponseCode = 4185
+	DockerContainerNotSpecified            ResponseCode = 4311
+	EmptyOrchestratorContext               ResponseCode = 4321
+	InvalidPrimaryIPConfig                 ResponseCode = 4341
```
Comment (nairashu) [inline cns/errorcodeframework/codes.go:27]: This error should be in the x100s as it is related to network configs.

File: cns/errorcodeframework/codes.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	StatusUnauthorized                     ResponseCode = 4561
+	UnexpectedError                        ResponseCode = 4580
+	UnsupportedAPI                         ResponseCode = 4591
+	AddressUnavailable                     ResponseCode = 5101
+	CallToHostFailed                       ResponseCode = 5105
+	InconsistentIPConfigState              ResponseCode = 5110
+	NetworkContainerPublishFailed          ResponseCode = 5133
+	NetworkContainerUnpublishFailed        ResponseCode = 5135
+	NetworkJoinFailed                      ResponseCode = 5138
+	NotFound                               ResponseCode = 5140
```
Comment (nairashu) [inline cns/errorcodeframework/codes.go:46]: What is the object NotFound error here? Is it ServerNotFound or NC Not Found Error?

File: cns/errorcodeframework/error_docs.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# Azure CNS HTTP Error Code Framework
```
Comment (nairashu) [inline cns/errorcodeframework/error_docs.md:1]: @rbtr : I do not remember what we finalized about having design docs in code in the acn repo. Is this file ok to be here or should we move it to Eng Hub or a wiki?

File: cns/errorcodeframework/error.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package errorcodeframework
```
Comment (nairashu) [inline cns/errorcodeframework/error.go:1]: I do like the suggestion provided by @ramiro-gamarra around the standard Error framework design across Azure Core. 

The structure: 
CNSResponse{
  "HTTPStatus": <Standard HTTP status code of success or failure>
"ResponseJson": 
  "error": {
    "code": "InvalidPasswordFormat",
    "message": "Human-readable description",
    "target": "target of error",
    "innererror": {
      "code": "PasswordTooShort",
      "minLength": 6,
    }
  }
}

**Review by nairashu — CHANGES_REQUESTED** [review]
Let's try to modify the framework to the standard pattern provided in Azure.

File: cns/errorcodeframework/error_docs.md (hunk 4 above)
Comment (rbtr) [reply cns/errorcodeframework/error_docs.md:1]: it is appropriate for this documentation to be here. it only talks about things in this opensource repo

**Pull request closed without merging**